db/migrate/20200727163200_change_comment_ad_events_github_code_ints_to_bigints.rb

```diff
@@ -0,0 +1,51 @@
+class ChangeCommentAdEventsGithubCodeIntsToBigints < ActiveRecord::Migration[6.0]
+  def up
+    ActiveRecord::Base.connection.execute("DROP VIEW IF EXISTS hypershield.display_ad_events")
```

Comment:
I'm curious: why do we need to explicitly `DROP VIEW` for the hypershield equivalent of the table? Does hypershield look to our DB and if we modify a column type, it blows up?

Which leads me to my second question: do we need to "recreate" the hypershield view for this table once we alter the column type?

These questions are really just me trying to understand how hypershield (blazer?) interacts with our DB and the order of operations here 😃

Reply:
2 GREAT questions

> I'm curious: why do we need to explicitly DROP VIEW for the hypershield equivalent of the table? 

I am not totally sure WHY but when I started doing these migrations I found that even tiny migrations would timeout with this error and if you scroll to the bottom of the backtrace you can see if goes back to hypershield removing the view. I also found when testing migrations on a fork of production that sometimes columns could not be updated bc they had hypershield views that relied on them. With these two things in mind I now drop the views before running the updates.
```
Caused by:
PG::QueryCanceled: ERROR:  canceling statement due to statement timeout
/app/vendor/bundle/ruby/2.7.0/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/postgresql/database_statements.rb:92:in `exec'
/app/vendor/bundle/ruby/2.7.0/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/postgresql/database_statements.rb:92:in `block (2 levels) in execute'
/app/vendor/bundle/ruby/2.7.0/gems/activesupport-6.0.3.2/lib/active_support/dependencies/interlock.rb:48:in `block in permit_concurrent_loads'
/app/vendor/bundle/ruby/2.7.0/gems/activesupport-6.0.3.2/lib/active_support/concurrency/share_lock.rb:187:in `yield_shares'
/app/vendor/bundle/ruby/2.7.0/gems/activesupport-6.0.3.2/lib/active_support/dependencies/interlock.rb:47:in `permit_concurrent_loads'
/app/vendor/bundle/ruby/2.7.0/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/postgresql/database_statements.rb:91:in `block in execute'
/app/vendor/bundle/ruby/2.7.0/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/abstract_adapter.rb:722:in `block (2 levels) in log'
/app/vendor/bundle/ruby/2.7.0/gems/activesupport-6.0.3.2/lib/active_support/concurrency/load_interlock_aware_monitor.rb:26:in `block (2 levels) in synchronize'
/app/vendor/bundle/ruby/2.7.0/gems/activesupport-6.0.3.2/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `handle_interrupt'
/app/vendor/bundle/ruby/2.7.0/gems/activesupport-6.0.3.2/lib/active_support/concurrency/load_interlock_aware_monitor.rb:25:in `block in synchronize'
/app/vendor/bundle/ruby/2.7.0/gems/activesupport-6.0.3.2/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `handle_interrupt'
/app/vendor/bundle/ruby/2.7.0/gems/activesupport-6.0.3.2/lib/active_support/concurrency/load_interlock_aware_monitor.rb:21:in `synchronize'
/app/vendor/bundle/ruby/2.7.0/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/abstract_adapter.rb:721:in `block in log'
/app/vendor/bundle/ruby/2.7.0/gems/activesupport-6.0.3.2/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
/app/vendor/bundle/ruby/2.7.0/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/abstract_adapter.rb:712:in `log'
/app/vendor/bundle/ruby/2.7.0/gems/activerecord-6.0.3.2/lib/active_record/connection_adapters/postgresql/database_statements.rb:90:in `execute'
/app/vendor/bundle/ruby/2.7.0/gems/hypershield-0.2.0/lib/hypershield.rb:146:in `execute'
/app/vendor/bundle/ruby/2.7.0/gems/hypershield-0.2.0/lib/hypershield.rb:27:in `block in drop_view'
/app/vendor/bundle/ruby/2.7.0/gems/hypershield-0.2.0/lib/hypershield.rb:26:in `each'
```

> Which leads me to my second question: do we need to "recreate" the hypershield view for this table once we alter the column type?

This is done automatically thanks to #9546. After we run the migrations which are kicked off by `db:prepare` then `hypershield:refresh` is run which will repopulate the views.

Reply:
This is super informative, thanks for the explanation, @mstruve ! 🙌

Reply:
I'm not 100% as to why it locks, but I believe it's because PostgreSQL needs to have the same exact data types as the underlying table, here we do change it from `int` to `bigint` so as the views are readonly they need to be recreated from scratch